src/GafferScene/PrimitiveSampler.cpp

```diff
@@ -0,0 +1,385 @@
+}
+
+Gaffer::StringPlug *PrimitiveSampler::prefixPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 3 );
+}
+
+const Gaffer::StringPlug *PrimitiveSampler::prefixPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 3 );
+}
+
+Gaffer::StringPlug *PrimitiveSampler::statusPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 4 );
+}
+
+const Gaffer::StringPlug *PrimitiveSampler::statusPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 4 );
+}
+
+bool PrimitiveSampler::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return
+		Deformer::affectsProcessedObject( input ) ||
+		input == sourcePlug() ||
+		input == sourceLocationPlug() ||
+		input == primitiveVariablesPlug() ||
+		input == prefixPlug() ||
+		input == statusPlug() ||
+		input == sourcePlug()->objectPlug() ||
+		input == inPlug()->transformPlug() ||
+		input == sourcePlug()->transformPlug() ||
+		affectsSamplingFunction( input )
+	;
+}
+
+void PrimitiveSampler::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	PrimitiveSampler::Deformer::hashProcessedObject( path, context, h );
+
+	const string sourceLocation = sourceLocationPlug()->getValue();
+	const string primitiveVariables = primitiveVariablesPlug()->getValue();
+	const string status = statusPlug()->getValue();
+	if( sourceLocation.empty() || ( primitiveVariables.empty() && status.empty() ) )
+	{
+		return;
+	}
+
+	ScenePlug::ScenePath sourcePath;
+	ScenePlug::stringToPath( sourceLocation, sourcePath );
+	if( !SceneAlgo::exists( sourcePlug(), sourcePath ) )
+	{
+		return;
+	}
+
+	h.append( sourcePlug()->objectHash( sourcePath ) );
+	h.append( primitiveVariables );
+	prefixPlug()->hash( h );
+	h.append( status );
+	h.append( inPlug()->fullTransformHash( path ) );
+	h.append( sourcePlug()->fullTransformHash( sourcePath ) );
+	hashSamplingFunction( h );
+}
+
+IECore::ConstObjectPtr PrimitiveSampler::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const Primitive *inputPrimitive = runTimeCast<const Primitive>( inputObject );
+	if( !inputPrimitive )
+	{
+		return inputObject;
+	}
+
+	const string sourceLocation = sourceLocationPlug()->getValue();
+	const string primitiveVariables = primitiveVariablesPlug()->getValue();
+	const string status = statusPlug()->getValue();
+	if( sourceLocation.empty() || ( primitiveVariables.empty() && status.empty() ) )
+	{
+		return inputObject;
+	}
+
+	PrimitiveVariable::Interpolation outputInterpolation = PrimitiveVariable::Invalid;
+	SamplingFunction samplingFunction = computeSamplingFunction( inputPrimitive, outputInterpolation );
+	if( !samplingFunction || outputInterpolation == PrimitiveVariable::Invalid )
+	{
+		return inputObject;
+	}
+
+	ScenePlug::ScenePath sourcePath;
+	ScenePlug::stringToPath( sourceLocation, sourcePath );
+	if( !SceneAlgo::exists( sourcePlug(), sourcePath ) )
+	{
+		return inputObject;
```

Comment:
Should this (and the next check) error too? Feels like if the user has specified a source location, and it doesn't exist, they might want to know?

Reply:
That is not our current convention (see CopyPrimitiveVariables), and whenever we've errored for this sort of thing we've received pushback and found people making strange workarounds. There may be a case for optional errors, but I'm reluctant to add complexity without a use case. I suspect the right approach might not be optional errors at all, but instead warnings that allow processing to continue but are flagged in the UI - that'd need to be a separate effort.